USB: Bluetooth H4 over bulk endpoints

Vudentz opened the pull request:
Initial code for support using H4 over bulk endpoints which has the following advantages over the standard H2:
 - Race free since everything is serialized.
 - Has support for ISO packets (0x05) in the standard

This can be used both with  a dedicated PID (hci_usb_h4) or with the existing hci_usb, for the later the host must enable it using the new vendor command Set Host Transport Mode (0x0010).

Vudentz committed:
USB: Add driver and PID for Bluetooth H4

This adds USB_PID_BLE_HCI_H4_SAMPLE along with it driver which uses H4
over bulk endpoints.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: hci_usb_h4: Add sample for USB_PID_BLE_HCI_H4_SAMPLE

This sample select CONFIG_USB_PID_BLE_HCI_H4_SAMPLE to use H4 over bulk
endpoints.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: hci_usb: Add implementation of Read/Set USB Transport Mode

This implements Read/Set USB Transport Mode in the Bluetooth class.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: hci: Add documentation for Read/Set USB Transport Mode

This adds documentation for Read/Set USB Transport Mode.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

jhedberg commented on samples/subsys/usb/usb_pid.Kconfig:
```diff
@@ -44,3 +44,7 @@ config USB_PID_WEBUSB_SAMPLE
+
+config USB_PID_BLE_HCI_H4_SAMPLE
+	hex
+	default 0x000C
```
Unrelated to this PR, but I had completely missed the existence of this kind of dependency from the USB subsystem to samples applications. Seems like it was added a year ago. Is this really the right way to go? It doesn't seem right to me that we would have sample app references from the core kernel or its subsystems. Why can't we just have per-sample "defaults" in each sample's prj.conf?

Samples are just that - samples. They act as a reference to the creation of final product applications, i.e. are mostly an ephemeral step in the app creation process. Having dedicated Kconfig options for them in the kernel or its subsystems feels therefore rather weird.

Vudentz replied:
I have similar thought, though we do need to track the PIDs in use for the samples otherwise it may confuse possible drivers if the samples end up reusing the same PID for multiple purposes, but like you said they are no suppose to be used in production, or at least we are not reserving this like that, anyway the application can overwrite this value if it want to use another PID, or perhaps we could have be using a single config option inform what PID shall be used then it is up to the sample itself to set it.

jhedberg replied:
I fully agree that we should coordinate (i.e. document) the PIDs that are used together with the Zephyr VID. That will help avoid conflicts when someone wants to do something custom with the Zephyr VID. I'd suggest we put this info somewhere in `doc/reference/usb`.

carlescufi commented:
Simple explanation for the USB folks that are not Bluetoot experts:
HCI has 3 basic types of data (ignoring audio):
- Commands (from Host to Controller)
- Events (from Controller to Host)
- Data (bidirectional)

In the Bluetooth spec, the following endpoints are defined:
- Control EP (OUT) for Commands
- Interrupt EP (IN) for Events
- Bulk (OUT) for data that goes Host->Controller
- Bulk (IN) for data that goes Controller->Host

What this new protocol tries to fix is the following race condition, summarized here as a simplified example of the encryption race:

Logical sequence:
```
Host->Controller CMD_Encryption_Start on the 
... Encryption happens over the air ...
Controller->Host: EVT_Encryption_Complete
... Some data is received over the air, with the link already encrypted ...
Controller->Host:  DATA_received_data
```
Real sequence:
```
Host sends CMD_Encryption_Start  on Control Endpoint (OUT)
Controller receives CMD_Encryption_Start, starts encryption over the air
... encryption completes over the air ...
... data is received over the air ..
Controller sends EVT_Encryption_Complete on Interrupt Endpoint (IN)
Controller sends DATA_received_data on Bulk Endpoint (IN)
Host receives DATA_received_data on Bulk Endpoint (IN)
Host receives EVT_Encryption_Complete on Interrupt Endpoint (IN)
Host thinks that the data has been received over an unencrypted link, which is not correct
```

jfischer-phytec-iot commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,325 @@
+
+	/* Check the USB status and do needed action if required */
+	switch (status) {
+	case USB_DC_ERROR:
```
You can remove unused cases if you like, I do not think that every class have to broadcast it :-)

Vudentz replied:
This is just used for logging purpose though but Im fine having a default that just print unhandled state or something like that.

jfischer-phytec-iot commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,325 @@
+	}
+
+	buf = net_buf_alloc(&tx_pool, K_FOREVER);
+	__ASSERT_NO_MSG(buf);
```
With ASSERT disabled that will be never caught, also not in net_buf_simple_reserve(). I know there is the same code in acl_read_cb, but why not check explicitly for NULL and log the error?

joerchan replied:
Or use BT_ASSERT, which are always enabled.

Vudentz replied:
Right, either this is no expected behavior since we are using k_FOREVER but I can use BT_ASSERT instead.

Vudentz replied:
Looks like bringing BT_ASSERT would possible require making it part of a public API so I will leave this for now and just enable asserts on the sample.

jfischer-phytec-iot replied:
I do not think it is acceptable from security perspective, either it has to be checked against NULL (`if (buf == NULL)`) or not (`__ASSERT_NO_MSG(buf);` not necessary at all) regardless of the build options/config.

Vudentz replied:
Alright then I would removed since using K_FOREVER should never really fail here.

jfischer-phytec-iot commented on samples/bluetooth/hci_usb_h4/README.rst:
```diff
@@ -0,0 +1,23 @@
+Requirements
+************
+
+* Bluetooth stack running on the host (e.g. BlueZ)
```
Does it depend on btusb driver/kernel-version?

Vudentz replied:
It does, @holtmann will be working on that.

jfischer-phytec-iot replied:
okay, please let usb people on the slack know when the btusb patches are coming.

Vudentz replied:
Sure, I will update this series when we are done with the btusb changes, we do have an out of the tree driver that works already.

Vudentz replied:
Yes there is the out of tree driver the @holtmann created:

https://github.com/holtmann/zpusb

joerchan replied:
I get "Page not found" on that link. But not important. :)

Vudentz replied:
Perhaps the tree is still not public, need to check with @holtmann

jfischer-phytec-iot commented on samples/bluetooth/hci_usb_h4/prj.conf:
```diff
@@ -0,0 +1,14 @@
+CONFIG_USB_DEVICE_STACK=y
+CONFIG_LOG=y
+CONFIG_USB_DRIVER_LOG_LEVEL_ERR=y
+CONFIG_USB_DEVICE_BT_H4=y
```
What do you think about to merge it with samples/bluetooth/hci_usb and add configuration overlays for USB_DEVICE_BLUETOOTH and USB_DEVICE_BT_H4?

Vudentz replied:
I think for USB_DEVICE_BT_H4 alone it was simpler, but I don't oppose to transform it on a overlay it just seems confusing since that would create several modes of operation to it.

Vudentz replied:
The hci_usb_h4 is not using the standard descriptors, it just has the bulk endpoints and which is why it uses a dedicated PID.

jfischer-phytec-iot commented on subsys/usb/class/Kconfig:
```diff
@@ -107,6 +107,21 @@ config BLUETOOTH_BULK_EP_MPS
+	int
+	depends on USB_DEVICE_BT_H4
+	default 64
+	range 8 64
```
This `range 8 64` is probably not necessary, is there any benefit to have smaller endpoint buffers?

Vudentz replied:
Nope, actually this might not work anymore since the buffer must be big enough to be used with a command.

emob-nordic replied:
Shouldn't we then make it hard coded? I know [bluetooth.c](https://github.com/zephyrproject-rtos/zephyr/blob/2cceb6220642928e08657e63736fab6cd5e0a6b7/subsys/usb/class/bluetooth.c#L92) has something similar, but is making this configurable helps anywhere?

Vudentz replied:
Right, there are other config option to setup the ACL length so I don't think we need yet another config option.

jfischer-no reviewed (commented):
LGTM, just few comments.

Vudentz committed:
Bluetooth: HCI: Add helpers to create events

This adds bt_hci_evt_create and bt_hci_cmd_complete functions.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
USB: Bluetooth: Remove BLUETOOTH_BULK_EP_MPS

Allowing to configuring the MPS different then the buffer size might
actually generate data fragmentation/reassembly so this just use
BT_BUF_ACL_SIZE which is set accourding to the controller buffer size.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
USB: Bluetooth: Remove BLUETOOTH_INT_EP_MPS

Allowing to configuring the MPS different then the buffer size might
actually generate data fragmentation/reassembly so this just use
BT_BUF_RX_SIZE which is set accourding to the controller buffer size.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,314 @@
+#define BT_H4_IN_EP_ADDR		0x81
+
+#define BT_H4_OUT_EP_IDX		0
+#define BT_H4_IN_EP_IDX			1
```
Indentation is off.

Vudentz replied:
It juts github that don't seem to format tab properly, will have to use space to indent.

joerchan replied:
I think it should be spaces anyway.

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -32,11 +33,19 @@ NET_BUF_POOL_DEFINE(tx_pool, CONFIG_BT_HCI_CMD_COUNT, CMD_BUF_SIZE,
 /* ACL data TX buffers */
+#if defined(CONFIG_USB_DEVICE_BLUETOOTH_VS_H4)
+#if defined(CONFIG_BT_CTLR_TX_BUFFERS)
+#define ACL_BUF_COUNT MAX(CONFIG_BT_HCI_CMD_COUNT, CONFIG_BT_CTLR_TX_BUFFERS)
```
I don't follow this. Why is it MAX in this expression.
If they are allocated from the same pool, shouldn't they be added?

Vudentz replied:
Yep, it be the sum of commands and acl buffers.

joerchan commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,314 @@
+/*
```
Is this the same as the class/bluetooth, except it does not have dynamic h4/h2 handling?

Vudentz replied:
This is using a dedicated PID so it doesn't use the startand bluetooth class.

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -137,6 +145,34 @@ static struct usb_ep_cfg_data bluetooth_ep_data[] = {
 	},
 };
 
+#define H4_CMD 0x01
```
Should we get a BT_H4 header file and move some defines in there, we redefine these all over now.
Maybe some functionality could be moved there as well? like h4_recv/send common parts?

Vudentz replied:
If there are more driver that would benefit perhaps we could do that, but for the initial driver I rather have something simpler so we don't have to start introducing too many public APIs until we settle on the details.

joerchan replied:
Sure, no need to do in this PR.

Vudentz replied:
Yes, will be done later on a separate PR.

emob-nordic replied:
This is unused after last push.

Vudentz commented on subsys/bluetooth/controller/hci/hci.c:
```diff
@@ -121,17 +121,33 @@ void hci_evt_create(struct net_buf *buf, u8_t evt, u8_t len)
-	struct bt_hci_evt_cmd_complete *cc;
+	struct net_buf *buf;
+
+	buf = bt_buf_get_evt(BT_HCI_EVT_CMD_COMPLETE, false, K_FOREVER);
```
Right, will fix this.

Vudentz commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -1014,6 +1015,37 @@ by the controller for the requested Handle_Type and Handle.
+Zephyr Set USB Transport Mode Command
+=====================================
+
+This command set the host transport mode and reset the controller.
```
Sure, I thought that would be obvious but I can spell out what reset it would be doing.

joerchan replied:
Thanks. 

Is the transport mode retained over regular HCI Reset command? It would be good to clarify the behavior for that as well. Not sure which option is best here.

Vudentz replied:
HCI reset resets just the HCI portion, H:2/H:4 is bellow that so it shouldn't be affected by it, we do reset in order to start fresh and don't have any pending events, etc, sending an HCI Reset won't affect the mode.

Vudentz commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -1014,6 +1015,37 @@ by the controller for the requested Handle_Type and Handle.
++--------------------------+-------+--------------------+--------------------+
+| Read_USB_Transport_Mode  | 0x010 |                    | Status             |
+|                          |       |                    | Num_Supported_Modes|
+|                          |       |                    | Supported_Mode[i]  |
```
The use of a list was recommended by @holtmann, my initial idea was to use bitmaps.

holtmann replied:
The list is more efficient and more extensible than a bit mask. Especially in cases where a bit mask then later has to be mapped back to a value.

joerchan replied:
> The list is more efficient and more extensible than a bit mask. Especially in cases where a bit mask then later has to be mapped back to a value.

Yeah, like the PHY bits, still bitmask is what is normally used so thought maybe we should stick with the consistency. No strong preference though.

Vudentz commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -42,7 +42,8 @@ command is supported.
 	|       | 6    | Read Tx Power Level (per Role/Connection)  |
-	|       | 7    | Reserved                                   |
+	|       | 7    | Get USB Mode                               |
+	|       | 7    | Set USB Mode                               |
```
Initially it was made independent of USB but @holtmann suggested it to be just for USB since so far is the only transport known to have races, the read was introduce more for the sake of completeness.

I will fix the bits though.

Vudentz commented on subsys/usb/class/bluetooth.c:
```diff
@@ -137,6 +145,34 @@ static struct usb_ep_cfg_data bluetooth_ep_data[] = {
+#define H4_SCO 0x03
+#define H4_EVT 0x04
+
+static void bt_h4_send(struct net_buf *buf)
```
It depends the way you look at it, I had the opposite meaning on purpose since bt_h4_send is internal to the usb driver the same way hci_rx_thread to receiving from the stack, but perhaps having it as usb_h4_send would have been clearer.

joerchan replied:
Yeah, I think usb_h4_send/recv would be clearer.

Vudentz committed:
USB: Bluetooth: Remove assert for NULL buffer

Buffer are allocated with K_FOREVER should never fail, in addition to
that asserts needs to be enabled and is not worth converting to runtime
check given the use of K_FOREVER.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented on subsys/usb/class/bluetooth.c:
```diff
@@ -66,6 +66,10 @@ NET_BUF_POOL_DEFINE(acl_rx_pool, ACL_BUF_COUNT, BT_BUF_ACL_SIZE,
 #define BLUETOOTH_OUT_EP_ADDR		0x02
 #define BLUETOOTH_IN_EP_ADDR		0x82
 
+/* TODO: Replace use of USB_MAX_FS_INT_MPS if higher speeds are supported */
```
This seems to be a limitation on USB stack, so we can't set values bigger than supported by the current speed as the stack would not truncate and expose the value as is to the host causing warning such as:

[50969.954964] usb 1-3: config 1 interface 0 altsetting 0 endpoint 0x1 has invalid maxpacket 107, setting to 64
[50969.954967] usb 1-3: config 1 interface 0 altsetting 0 endpoint 0x82 has invalid maxpacket 107, setting to 64

Vudentz replied:
Afaik fragmentation is already handled by HCI layer so we don't have to do in USB again.

Vudentz replied:
Let me take this back, it does seem that USB itself would need to handle fragmentation somehow because the HCI would fragment accourding to the HCI buffer size not the USB MPS, @jhedberg @holtmann or do we actually break the fragments accourding to the MPS on the drivers? In any case there doesn't seem to be any header that would make reassemble possible, so we can't really tell if that is a fragment of a full packet, beside on TX path we don't actually attempt to fragment either so I assume the USB core would have to deal with that. We could also force disable the fragmentation at USB by forcing HCI buffer to be <= 64.

joerchan replied:
There is BLUETOOTH CORE SPECIFICATION Version 5.2 | Vol 4, Part B which talks about USB frames and sizes. For now I think we can postpone this for this PR. Limiting the HCI buffer to be <= 64. Which would mean limiting the controller TX/RX buffer, CONFIG_BT_CTLR_TX_BUFFERS/CONFIG_BT_CTLR_RX_BUFFERS. It would limiting us for data length extension and advertising extension for the time being.

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -251,6 +326,91 @@ static void bluetooth_status_cb(struct usb_cfg_data *cfg,
+	cc = net_buf_add(buf, sizeof(*cc));
+	cc->status = status;
+
+	net_buf_put(&rx_queue, buf);
```
Should this be `tx_queue`?

Vudentz replied:
Yep

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -251,6 +326,91 @@ static void bluetooth_status_cb(struct usb_cfg_data *cfg,
+	net_buf_add_u8(rsp, BT_HCI_VS_USB_H2_MODE);
+	net_buf_add_u8(rsp, BT_HCI_VS_USB_H4_MODE);
+
+	net_buf_put(&rx_queue, rsp);
```
should this be `tx_queue`?

Vudentz replied:
Yep, forgot to change this one.

joerchan commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,315 @@
+
+	switch (bt_buf_get_type(buf)) {
+	case BT_BUF_ACL_IN:
+		net_buf_push_u8(buf, H4_ACL);
```
This is the BT_BUF_RESERVE, Right? This is now coming from the sample prj.conf. Could we change this dependency?

Vudentz replied:
Yes, I might change it to use select.

joerchan commented on include/bluetooth/hci.h:
```diff
@@ -1825,6 +1825,35 @@ typedef bool bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+
+/** Allocate an HCI event buffer.
+ *
+ * This function allocates a new buffer for a HCI event. It is given the
```
text: "an HCI".

Vudentz replied:
Actually it should a HCI, since we do pronounce H and even if that was not the case it followed by a consonant not a vogal: https://editingandwritingservices.com/a-or-an-before-words-beginning-with-h/

joerchan replied:
When it comes to abbreviations it is the pronunciation that controls it, for example an FBI agent. I think "an HCI event" ("an aitch si ai event") is correct.

Vudentz replied:
I guess that refers to the second term then, an event/an agent, but it wouldn't work for HCI command since that is a command not an command.

joerchan commented on include/bluetooth/hci.h:
```diff
@@ -1825,6 +1825,35 @@ typedef bool bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+ */
+struct net_buf *bt_hci_evt_create(u8_t evt, u8_t len);
+
+/** Allocate a HCI Command Complete event buffer.
```
text: "an HCI".

joerchan commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,315 @@
+#define BT_H4_BULK_EP_MPS               MIN(BUF_SIZE, USB_MAX_FS_BULK_MPS)
+
+/* HCI RX/TX threads */
+static K_THREAD_STACK_DEFINE(rx_thread_stack, 512);
```
Should these come from Kconfig?

Vudentz replied:
Don't think one would need to change the stack size, that is why there are dedicated threads instead of using something like the syswq.

joerchan replied:
The RX thread will call bt_send which will call `bt_dev.drv->send`. We have the possebility to register different hci_drivers here, and those drivers might be (as in my case) out of tree drivers.

Vudentz replied:
How it is working now then? It doesn't seem to be configurable in bluetooth.c either, why would I  have to assume this would be different to bt_h4, if you have a requirement not attended by the stack please send a patch since it is kind counter productive for me to guess how your out of the tree driver would work I may not even guess what default and range would work best which is another reason to have this pushed as soon as possible so I don't have to resolve every driver problem in one, never ending, PR. For the most part this is just an example for how H4 mode would work as stand alone mode, don't think anyone would be thinking on productize it without it being part of the Bluetooth spec.

joerchan replied:
Right. Then it doesn't work in that case for bluetooth.c either. I'll make a note of it and and fix it for our out of tree driver afterwards if needed. So just leave it as is.

joerchan commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,315 @@
+	}
+
+	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
```
Does the read USB transport modes need to be handled for this driver?
For a USB host that support both it could read the supported command, and if it sees only h4 is supported in knows it doesn't need to change the mode.

Vudentz replied:
This does not use the standard Bluetooth class, so the driver would need to know it is already in H:4 mode, in fact even the Read USB Transport would fail if is sent without the header so the driver must know it only works with H:4 mode so that command is not very useful here.

Vudentz commented on subsys/usb/class/bluetooth.c:
```diff
@@ -45,19 +54,29 @@ NET_BUF_POOL_DEFINE(tx_pool, CONFIG_BT_HCI_CMD_COUNT, CMD_BUF_SIZE,
 
 /* Data size needed for ACL buffers */
+#if defined(CONFIG_USB_DEVICE_BLUETOOTH_VS_H4)
+#define BT_BUF_ACL_SIZE (BT_BUF_RX_SIZE + BT_L2CAP_BUF_SIZE(BT_L2CAP_MTU))
```
Yep, this is the actual buffer size not the count.

Vudentz commented on samples/bluetooth/hci_usb_h4/prj.conf:
```diff
@@ -0,0 +1,15 @@
+
+CONFIG_BT=y
+CONFIG_BT_HCI_RAW=y
+CONFIG_BT_HCI_RAW_RESERVE=1
```
Meaning select with CONFIG_USB_DEVICE_BT_H4, I guess it could though we haven't done that for the standard class for some reason.

Vudentz commented on subsys/usb/class/bluetooth.c:
```diff
@@ -274,6 +434,40 @@ static int bluetooth_class_handler(struct usb_setup_packet *setup,
+		net_buf_simple_restore(&buf->b, &state);
+	}
+
 	net_buf_put(&tx_queue, buf);
```
Yes, now rx and tx are relative to USB not Bluetooth, so you are right that this should be put into the rx_queu.

Vudentz commented on include/bluetooth/hci.h:
```diff
@@ -1825,6 +1825,35 @@ typedef bool bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+ *
+ * @return Newly allocated buffer.
+ */
+struct net_buf *bt_hci_cmd_complete(u16_t op, u8_t plen);
```
Will add, it looks like there is a lot code that just set the status and nothing else, though if there is more than just the status bt_hci_cmd_complete should be used.

Vudentz replied:
Hmm I guess you are refering to, bt_hci_evt_cmd_status rather than bt_hci_evt_cc_status, anyway will add bt_hci_cmd_status.

jhedberg commented on include/bluetooth/hci.h:
```diff
@@ -1825,6 +1825,49 @@ typedef bool bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+/** Allocate a HCI event buffer.
+ *
+ * This function allocates a new buffer for a HCI event. It is given the
+ * OpCode (encoded e.g. using the BT_OP macro) and the total length of the
```
BT_OP is used for command opcodes not for event codes. I think it'd be good to disambiguate here and use "event code" events.

Vudentz replied:
Right, will fix that.

jhedberg commented on include/bluetooth/hci.h:
```diff
@@ -1825,6 +1825,49 @@ typedef bool bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+ *
+ * @return Newly allocated buffer.
+ */
+struct net_buf *bt_hci_cmd_complete(u16_t op, u8_t plen);
```
Please be consistent with the choice of verb since you're making these public APIs. You have "create" with `bt_hci_evt_create()` but this one is missing a verb completely. Same goes for `bt_hci_cmd_status()`.

I'm a little bit doubtful whether we need these at all TBH since they're so simple (i.e. users could just use the bt_buf APIs directly instead).

Vudentz replied:
They would be duplicated though, does that bt_hci_cmd_complete_create makes it sounds any better?

joerchan replied:
I think `bt_hci_create_cmd_complete` is the naming convention, but that would mean `bt_hci_evt_create` is not following the same naming convention.

Vudentz replied:
No bt_hci_cmd_create, bt_hci_cmd_snd, etc, the verb always comes last so the convention is the other way around at least when it comes to hci_raw.

emob-nordic commented on subsys/usb/class/bluetooth.c:
```diff
@@ -168,14 +215,43 @@ static void hci_rx_thread(void)
+		LOG_ERR("Error sending to driver");
+		net_buf_unref(buf);
+	}
+}
```
This is similar to ```usb_h4_send```. I prefer one of the solitions:
- rename it to something like ```h4_cmd_type_update```. This is what actually happens there. And leave ```bt_send``` untouched in ```hci_rx_thread```
```diff
diff --git a/subsys/usb/class/bluetooth.c b/subsys/usb/class/bluetooth.c
index 81fe773834..f1e1bdd64a 100644
--- a/subsys/usb/class/bluetooth.c
+++ b/subsys/usb/class/bluetooth.c
@@ -215,7 +215,7 @@ static void hci_tx_thread(void)
 	}
 }
 
-static void usb_h4_recv(struct net_buf *buf)
+static void usb_h4_cmd_type_update(struct net_buf *buf)
 {
 	u8_t type;
 
@@ -234,11 +234,6 @@ static void usb_h4_recv(struct net_buf *buf)
 	}
 
 	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
-
-	if (bt_send(buf)) {
-		LOG_ERR("Error sending to driver");
-		net_buf_unref(buf);
-	}
 }
 
 static void hci_rx_thread(void)
@@ -250,8 +245,10 @@ static void hci_rx_thread(void)
 
 		if (IS_ENABLED(CONFIG_USB_DEVICE_BLUETOOTH_VS_H4) &&
 		    mode == BT_HCI_VS_USB_H4_MODE) {
-			usb_h4_recv(buf);
-		} else if (bt_send(buf)) {
+			usb_h4_cmd_type_update(buf);
+		}
+
+		if (bt_send(buf)) {
 			LOG_ERR("Error sending to driver");
 			net_buf_unref(buf);
 		}
```
- put this code into ```hci_rx_thread``` directly. It only changes the command type and later calls ```bt_send``` exactly the same way the H2 does.
``` diff
diff --git a/subsys/usb/class/bluetooth.c b/subsys/usb/class/bluetooth.c
index 81fe773834..e31264bc0d 100644
--- a/subsys/usb/class/bluetooth.c
+++ b/subsys/usb/class/bluetooth.c
@@ -215,32 +215,6 @@ static void hci_tx_thread(void)
 	}
 }
 
-static void usb_h4_recv(struct net_buf *buf)
-{
-	u8_t type;
-
-	type = net_buf_pull_u8(buf);
-
-	switch (type) {
-	case H4_CMD:
-		bt_buf_set_type(buf, BT_BUF_CMD);
-		break;
-	case H4_ACL:
-		bt_buf_set_type(buf, BT_BUF_ACL_OUT);
-		break;
-	default:
-		LOG_ERR("Unknown H4 type %u", type);
-		return;
-	}
-
-	LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
-
-	if (bt_send(buf)) {
-		LOG_ERR("Error sending to driver");
-		net_buf_unref(buf);
-	}
-}
-
 static void hci_rx_thread(void)
 {
 	while (true) {
@@ -250,8 +224,22 @@ static void hci_rx_thread(void)
 
 		if (IS_ENABLED(CONFIG_USB_DEVICE_BLUETOOTH_VS_H4) &&
 		    mode == BT_HCI_VS_USB_H4_MODE) {
-			usb_h4_recv(buf);
-		} else if (bt_send(buf)) {
+			switch (net_buf_pull_u8(buf)) {
+			case H4_CMD:
+				bt_buf_set_type(buf, BT_BUF_CMD);
+				break;
+			case H4_ACL:
+				bt_buf_set_type(buf, BT_BUF_ACL_OUT);
+				break;
+			default:
+				LOG_ERR("Unknown H4 type %u", type);
+				return;
+			}
+		}
+
+		LOG_DBG("buf %p type %u len %u", buf, bt_buf_get_type(buf), buf->len);
+
+		if (bt_send(buf)) {
 			LOG_ERR("Error sending to driver");
 			net_buf_unref(buf);
 		}
```


Same comment applies for ```usb_h4_recv``` and ```hci_rx_thread``` in bt_h4.c

Vudentz replied:
Right, so when I have to replace usb_h4_recv I would have to remove these all over the place, no thanks.

emob-nordic commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,315 @@
+#define H4_CMD 0x01
+#define H4_ACL 0x02
+#define H4_SCO 0x03
+#define H4_EVT 0x04
```
Should end up in header file (see review comment).

emob-nordic commented on subsys/usb/class/bt_h4.c:
```diff
@@ -0,0 +1,315 @@
+
+	usb_transfer_sync(bt_h4_ep_data[BT_H4_IN_EP_IDX].ep_addr,
+			  buf->data, buf->len, USB_TRANS_WRITE);
+
```
Same concerns as in bluetooth.c ```usb_h4_send```

emob-nordic commented on samples/bluetooth/hci_usb_h4/src/main.c:
```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
```
Shouldn't be 2020?

Vudentz replied:
Fixed.

eobalski reviewed (dismissed):
General notes:
- This PR changes tx_queue/rx_queue to determine direction from perspective of USB.
  Maybe  its worhwhile to change tx_queue/rx_queue to usb_tx_queue/usb_rx_queue.
  I think this may help for future maintaining the code.
- Some definitions are duplicated in bluetooth.c and bt_h4.c.
  I prefer to see common things in one place. I suggest to create usb_bluetooth.h.
  Furthermore I would put everything to ./class/bluetooth/ dir.
  Maybe its high time for those cleanups as well.
- Question: I am not sure. bluetooth h2/h4 have dedicated PID, right?
  I this case if H2 will get H2 PID how will host determine if this could be possibly changed to h4?
  I see there is read_usb_transport_mode but still host will 'quess' if should ask or it always asks for it?

Vudentz commented on subsys/usb/class/bluetooth.c:
```diff
@@ -137,14 +150,48 @@ static struct usb_ep_cfg_data bluetooth_ep_data[] = {
+		    mode == BT_HCI_VS_USB_H4_MODE) {
+			usb_h4_send(buf);
+			continue;
+		}
```
I will have to desagree, we will be moving to common usb_h4_send later on, you even suggested that with moving to a common file.

Vudentz committed:
Bluetooth: hci_raw: Add support for using H:4 transport

This adds 2 config options which enables hci_raw to work in
H:4 mode and enable it by default automatically.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -137,14 +150,25 @@ static struct usb_ep_cfg_data bluetooth_ep_data[] = {
+		buf = net_buf_get(&tx_queue, K_FOREVER);
+
+		if (IS_ENABLED(CONFIG_USB_DEVICE_BLUETOOTH_VS_H4) &&
+		    mode == BT_RAW_MODE_H4) {
```
Maybe add `bt_raw_get_mode`?

Vudentz replied:
Good idea, will do that.

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -263,7 +423,7 @@ static int bluetooth_class_handler(struct usb_setup_packet *setup,
 		return -EINVAL;
```
If H4 mode is enabled shouldn't this check BT_BUF_ACL_SIZE instead?

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -251,6 +276,87 @@ static void bluetooth_status_cb(struct usb_cfg_data *cfg,
+	struct bt_hci_rp_vs_read_usb_transport_mode *rp;
+
+	rsp = bt_hci_cmd_complete_create(BT_HCI_OP_VS_READ_USB_TRANSPORT_MODE,
+				  sizeof(*rp) + 2);
```
nit: alignment.

Vudentz replied:
Fixed.

joerchan commented on subsys/usb/class/bluetooth.c:
```diff
@@ -251,6 +276,87 @@ static void bluetooth_status_cb(struct usb_cfg_data *cfg,
+	net_buf_put(&tx_queue, buf);
+}
+
 static int bluetooth_class_handler(struct usb_setup_packet *setup,
```
I don't understand when this function is called. Will it still be called after switching to H4 mode? If not then it would seem that the vs_cmd handling will not be called anymore.

emob-nordic replied:
This would be called by USB stack if host invokes [class request](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/usb/usb_device.c#L1210)

Vudentz replied:
Right, if a vs command is sent over the bulk endpoint then this wouldn't so we have to handle both when receiving over the class request and when it sent over the bulk endpoint in case of H4 mode, so that needs fixing.

Vudentz commented on include/bluetooth/hci_raw.h:
```diff
@@ -32,6 +32,24 @@ extern "C" {
 int bt_send(struct net_buf *buf);
 
+enum {
+	/* No header mode. */
```
I was thinking on naming it PASSTHROUGH instead of NO_HEADER, that should be a little clearer by itself but I will drop a line of comment on the actual expected behavior.

Vudentz commented on subsys/usb/class/bluetooth.c:
```diff
@@ -137,14 +150,48 @@ static struct usb_ep_cfg_data bluetooth_ep_data[] = {
+	}
+
+	usb_transfer_sync(bluetooth_ep_data[HCI_IN_EP_IDX].ep_addr, buf->data,
+			  buf->len, USB_TRANS_WRITE);
```
This is using a bulk endpoint though which hasn't been ZPL flag which I guess it on purpose.

Vudentz committed:
Bluetooth: hci_raw: Add support for command extention

This adds support for registering a command extention table which is
used to match incoming commands and then pass the buffer to its
function handler.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

joerchan commented on include/bluetooth/hci_raw.h:
```diff
@@ -32,7 +32,86 @@ extern "C" {
+	 *
+	 *  @param buf Buffer containing the command.
+	 *
+	 *  @return HCI Status code or BT_HCI_ERR_EXT_HANDLED if command has
```
Maybe add note about returning `0` to send the command for  handling in the controller?

Vudentz replied:
Done.

joerchan reviewed (approved):
LGTM.

I would add a note about the ext-cmd handling 0xff vs 0x00 in the bt_raw header.

Vudentz commented:
So can we finally get this set merged?

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -71,10 +85,38 @@ struct net_buf *bt_buf_get_evt(u8_t evt, bool discardable, s32_t timeout)
+
+		if (err) {
+			return err;
+		}
```
If you're not planning to do anything special with `err` (like logging it) I'd just shorten this to `return bt_h4_recv(buf);`

Vudentz replied:
Actually we can't just return bt_h4_recv since that only sets the buffer type so we need to continue to send the buffer to the monitor, etc.

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -88,19 +130,162 @@ int bt_recv_prio(struct net_buf *buf)
+
+		if (err) {
+			return err;
+		}
```
If you're not planning to do anything special with `err` (like logging it) I'd just shorten this to `return bt_h4_send(buf);`

Vudentz replied:
Same reason as above.

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -88,19 +130,162 @@ int bt_recv_prio(struct net_buf *buf)
+	u8_t status;
+
+	if (!cmd_ext) {
+		return status;
```
huh? You're returning an uninitialised variable here.

Vudentz replied:
Fixed

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -88,19 +130,162 @@ int bt_recv_prio(struct net_buf *buf)
+		status = bt_send_ext(buf);
+		if (status) {
+			return 0;
+		}
```
A code comment or debug log would be appropriate here. So `bt_send_ext()` is failing but you're treating it as success. Why?

Vudentz replied:
It should have been the status actually, this is now fixed.

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -88,19 +130,162 @@ int bt_recv_prio(struct net_buf *buf)
+	if (IS_ENABLED(CONFIG_BT_HCI_RAW_H4)) {
+		switch (mode) {
+		case BT_RAW_MODE_PASSTHROUGH:
+		case BT_RAW_MODE_H4:
```
Why is "HCI" omitted from `BT_RAW_MODE_*`? Everything here centres around HCI so I think it'd be appropriate and consistent to have it there.

Vudentz replied:
Fixed.

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -88,19 +130,162 @@ int bt_recv_prio(struct net_buf *buf)
+	return -EINVAL;
+}
+
+u8_t bt_raw_get_mode(void)
```
Same here.

Vudentz replied:
Fixed.

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -88,19 +130,162 @@ int bt_recv_prio(struct net_buf *buf)
+	return BT_RAW_MODE_PASSTHROUGH;
+}
+
+void bt_raw_cmd_ext_register(struct bt_raw_cmd_ext *cmds, size_t size)
```
And here.

Vudentz replied:
Fixed.

jhedberg commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -1014,6 +1015,47 @@ by the controller for the requested Handle_Type and Handle.
+	| Value              | Parameter Description                |
+	+--------------------+--------------------------------------+
+	| 0x00               | H:2 transport mode                   |
+	| 0x01               | H:4 transport mode                   |
```
I think we need a proper definition of these modes. I know H:2 and H:4 are used colloquially amongst Bluetooth veterans, but those are references to ancient Bluetooth spec versions and sections in them that do not exist anymore. Either we should refer to sections in the latest core spec or describe the protocol format in detail if such sections do not exist.

joerchan replied:
How about "USB Mode" and "Serial Mode", H4 is currently used for SPI/UART/RPmsg.

Vudentz replied:
Ive added some description to each mode.

jhedberg commented on include/bluetooth/hci.h:
```diff
@@ -1856,6 +1856,48 @@ typedef bool bt_hci_vnd_evt_cb_t(struct net_buf_simple *buf);
+ *
+ * @return Newly allocated buffer.
+ */
+struct net_buf *bt_hci_evt_create(u8_t evt, u8_t len);
```
Are we sure `include/bluetooth/hci.h` is the most appropriate place for these? Wouldn't it always be the HCI driver that's the user of these, which would then make `include/drivers/bluetooth/hci_driver.h` seem like the more appropriate place?

Vudentz replied:
Sure, I will move them there.

jhedberg commented on include/bluetooth/hci_raw.h:
```diff
@@ -32,7 +32,88 @@ extern "C" {
+	 *  according to the buffer type when coming from the stack and will be
+	 *  removed and used to set the buffer type.
+	 */
+	BT_RAW_MODE_H4 = 0x01,
```
These seem to be missing "HCI" between "BT" and "RAW".

Vudentz replied:
Fixed.

jhedberg reviewed (dismissed):
> So can we finally get this set merged?

@Vudentz unfortunately it doesn't seem so, see my inline comments.

Also, didn't we agree to have the most obvious H:4 HCI raw user, i.e. `samples/bluetooth/hci_uart` converted as well? In fact, I'd have preferred if you just left all USB stuff out for now and only introduced the new H:4 HCI raw API together with the conversion of the `hci_uart` sample.

Not that I consider it a blocker for merging this, but more out of curiosity: has @holtmann given any feedback to the HCI vendor command additions?

jhedberg commented:
@Vudentz btw, if you do consider this ready for merging, please remove the RFC label, that's in some ways the same as DNM

holtmann commented:
> > So can we finally get this set merged?
> 
> @Vudentz unfortunately it doesn't seem so, see my inline comments.
> 
> Also, didn't we agree to have the most obvious H:4 HCI raw user, i.e. `samples/bluetooth/hci_uart` converted as well? In fact, I'd have preferred if you just left all USB stuff out for now and only introduced the new H:4 HCI raw API together with the conversion of the `hci_uart` sample.
> 
> Not that I consider it a blocker for merging this, but more out of curiosity: has @holtmann given any feedback to the HCI vendor command additions?

I gave feedback via IRC on the HCI commands, not sure if all changes are incorporated yet. There were a few minor issues.

Vudentz commented:
> > So can we finally get this set merged?
> 
> @Vudentz unfortunately it doesn't seem so, see my inline comments.
> 
> Also, didn't we agree to have the most obvious H:4 HCI raw user, i.e. `samples/bluetooth/hci_uart` converted as well? In fact, I'd have preferred if you just left all USB stuff out for now and only introduced the new H:4 HCI raw API together with the conversion of the `hci_uart` sample.

I looked into that, the problem there is that it actually using the header to discard data, etc, I was actually thinking we could unify this code around TX/RX threads so that hci_raw actually manages things like buffers and queuing by itself, but I wonder if wouldn't be better to merge these first so people can start experiementing with their own drivers.

holtmann commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -42,7 +42,8 @@ command is supported.
 	|       | 6    | Read Tx Power Level (per Role/Connection)  |
-	|       | 7    | Reserved                                   |
+	|       | 7    | Read USB Mode                              |
+	|       | 8    | Set USB Mode                               |
```
These are per octet blocks and Set USB Mode needs to go into octet 2.

Vudentz replied:
Right, I missed that.

holtmann commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -42,7 +42,8 @@ command is supported.
 	|       | 5    | Write Tx Power Level (per Role/Connection) |
 	|       | 6    | Read Tx Power Level (per Role/Connection)  |
-	|       | 7    | Reserved                                   |
+	|       | 7    | Read USB Mode                              |
```
Read Supported USB Transport Modes and Set USB Transport Mode

Vudentz replied:
Fixed.

holtmann commented on doc/reference/bluetooth/hci.txt:
```diff
@@ -1014,6 +1015,52 @@ by the controller for the requested Handle_Type and Handle.
 When the Read_Tx_Power_Level command has completed, a Command Complete
 event shall be generated.
 
+Zephyr Read USB Transport Mode Command
```
Zephyr Read Supported USB Transport Modes

Vudentz replied:
Fixed.

Vudentz commented:
@jhedberg @holtmann your comments have addressed.

Vudentz commented:
@jfischer-phytec-iot it seems there are no more coments to address.

Vudentz committed:
Bluetooth: hci_raw: Move buffer management to common place

This makes hci_raw to manage RX and TX buffers so its logic don't have
to be replicated on each an every driver/application, it also makes it
simpler to deal with extra headers for H:4 mode since that then can be
done at earlier at buffer allocation.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: hci_uart: Make use of CONFIG_BT_HCI_RAW_H4

This selects CONFIG_BT_HCI_RAW_H4 which indicates to hci_raw to handle
H:4 header automatically.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
So Ive updated this set to move the buffer management to hci_raw and then coverted hci_uart to use BT_HCI_RAW_H4 but it appear that the board I was testing with no longer works (nrf52840dk_nrf52840) anymore, not sure if this was caused by some renaming that did happen but master is also not working for me @carlescufi @joerchan

carlescufi commented:
> So Ive updated this set to move the buffer management to hci_raw and then coverted hci_uart to use BT_HCI_RAW_H4 but it appear that the board I was testing with no longer works (nrf52840dk_nrf52840) anymore, not sure if this was caused by some renaming that did happen but master is also not working for me @carlescufi @joerchan

How exactly doesn't it work anymore @Vudentz ? This should work as usual, even though the board was indeed renamed from nrf52840_pca10056.

carlescufi commented:
> > > So Ive updated this set to move the buffer management to hci_raw and then coverted hci_uart to use BT_HCI_RAW_H4 but it appear that the board I was testing with no longer works (nrf52840dk_nrf52840) anymore, not sure if this was caused by some renaming that did happen but master is also not working for me @carlescufi @joerchan
> > 
> > 
> > How exactly doesn't it work anymore @Vudentz ? This should work as usual, even though the board was indeed renamed from nrf52840_pca10056.
> 
> After flashing the device doesn't come back:
> 
> [1523445.247166] usb 1-3: new full-speed USB device number 82 using xhci_hcd
> [1523445.376032] usb 1-3: New USB device found, idVendor=1915, idProduct=521f, bcdDevice= 1.00
> [1523445.376037] usb 1-3: New USB device strings: Mfr=1, Product=2, SerialNumber=3
> [1523445.376041] usb 1-3: Product: Open DFU Bootloader
> [1523445.376044] usb 1-3: Manufacturer: Nordic Semiconductor
> [1523445.376077] usb 1-3: SerialNumber: E60487035F81
> [1523445.377907] cdc_acm 1-3:1.0: ttyACM0: USB ACM device
> [1523480.308657] cdc_acm 1-3:1.0: failed to set dtr/rts
> [1523480.308810] usb 1-3: USB disconnect, device number 82
> 
> I did try with master as well and had the same result.

OK, thanks. Does the standard CDC ACM sample work for you on this board? @emob-nordic can you try to reproduce?

emob-nordic commented:
> So Ive updated this set to move the buffer management to hci_raw and then coverted hci_uart to use BT_HCI_RAW_H4

Have You just flashed the hci_uart sample from this branch? I am a bit confused for what Carles asked me there.

Vudentz commented:
> > So Ive updated this set to move the buffer management to hci_raw and then coverted hci_uart to use BT_HCI_RAW_H4
> 
> Have You just flashed the hci_uart sample from this branch? I am a bit confused for what Carles asked me there.

Ive tried flashing hci_usb from master and it doesn't seems to work either, I don't have a board to test hci_uart, or perhaps I don't which board would be able to use it.

emob-nordic commented:
> ...
> Ive tried flashing hci_usb from master and it doesn't seems to work either, I don't have a board to test hci_uart, or perhaps I don't which board would be able to use it.

For me it works and I have no problems to run form master and this branch as well, dunno how I can help there...

Vudentz commented:
> > ...
> > Ive tried flashing hci_usb from master and it doesn't seems to work either, I don't have a board to test hci_uart, or perhaps I don't which board would be able to use it.
> 
> For me it works and I have no problems to run form master and this branch as well, dunno how I can help there...

Maybe something wrong with my board then, can you confirm that you got an HCI Controller with these changes? Which board are you testing with btw.

emob-nordic commented:
> > > ...
> > > Ive tried flashing hci_usb from master and it doesn't seems to work either, I don't have a board to test hci_uart, or perhaps I don't which board would be able to use it.
> > 
> > 
> > For me it works and I have no problems to run form master and this branch as well, dunno how I can help there...
> 
> Maybe something wrong with my board then, can you confirm that you got an HCI Controller with these changes? Which board are you testing with btw.

```
user@nordic:~/zephyr/zephyr/samples/bluetooth/hci_usb(master)$git log --oneline -3
5a149dfca6 (HEAD -> master, upstream/master) drivers: serial: uart_mcux: Convert to DT_INST
242b13cc8d boards: arm: nxp: convert to DT_NODELABEL usage for UART_MCUX
9f0aa69843 Z_ISR_DECLARE: update for C++ support
user@nordic:~/zephyr/zephyr/samples/bluetooth/hci_usb(master)$hcitool dev
Devices:
	hci1	50:EB:71:E7:D6:59
user@nordic:~/zephyr/zephyr/samples/bluetooth/hci_usb(master)$west flash > a
user@nordic:~/zephyr/zephyr/samples/bluetooth/hci_usb(master)$hcitool dev
Devices:
	hci0	00:00:00:00:00:00
	hci1	50:EB:71:E7:D6:59
user@nordic:~/zephyr/zephyr/samples/bluetooth/hci_usb(master)$
```

```
[ 9140.555319] usb 3-3.1: Product: USB-DEV
[ 9140.555320] usb 3-3.1: Manufacturer: ZEPHYR
[ 9140.555320] usb 3-3.1: SerialNumber: DB12E1CA912845B6
[ 9146.860154] usb 3-2: USB disconnect, device number 3
[ 9146.860234] cdc_acm 3-2:1.0: failed to set dtr/rts
[ 9161.047469] usb 3-3.1: USB disconnect, device number 12
[ 9182.142213] usb 3-2: new full-speed USB device number 13 using xhci_hcd
[ 9182.395669] usb 3-2: New USB device found, idVendor=2fe3, idProduct=000b
[ 9182.395670] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 9182.395671] usb 3-2: Product: USB-DEV
[ 9182.395672] usb 3-2: Manufacturer: ZEPHYR
[ 9182.395690] usb 3-2: SerialNumber: DB12E1CA912845B6
[ 9182.726257] usb 3-4.1: new high-speed USB device number 14 using xhci_hcd
[ 9182.935735] usb 3-4.1: New USB device found, idVendor=1366, idProduct=0105
[ 9182.935741] usb 3-4.1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 9182.935744] usb 3-4.1: Product: J-Link
[ 9182.935747] usb 3-4.1: Manufacturer: SEGGER
[ 9182.935750] usb 3-4.1: SerialNumber: 000683915891
[ 9182.944202] cdc_acm 3-4.1:1.0: ttyACM0: USB ACM device
[ 9188.968606] usb 3-2: USB disconnect, device number 13
[ 9211.542315] usb 3-2: new full-speed USB device number 15 using xhci_hcd
[ 9211.795660] usb 3-2: New USB device found, idVendor=2fe3, idProduct=000b
[ 9211.795663] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 9211.795665] usb 3-2: Product: USB-DEV
[ 9211.795666] usb 3-2: Manufacturer: ZEPHYR
[ 9211.795667] usb 3-2: SerialNumber: DB12E1CA912845B6
[ 9268.221979] usb 3-2: USB disconnect, device number 15
[ 9292.910385] usb 3-2: new full-speed USB device number 16 using xhci_hcd
[ 9293.161728] usb 3-2: New USB device found, idVendor=2fe3, idProduct=000b
[ 9293.161729] usb 3-2: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 9293.161730] usb 3-2: Product: USB-DEV
[ 9293.161730] usb 3-2: Manufacturer: ZEPHYR
[ 9293.161731] usb 3-2: SerialNumber: DB12E1CA912845B6
```

I've checked with nrf52840dk_nrf52840 (old naming: nrf52840_pca10056) on master branch (see above).

jfischer-phytec-iot commented on doc/reference/usb/index.rst:
```diff
@@ -38,6 +38,7 @@ The following Product IDs are currently used:
 * :option:`CONFIG_USB_PID_MASS_SAMPLE`
 * :option:`CONFIG_USB_PID_TESTUSB_SAMPLE`
 * :option:`CONFIG_USB_PID_WEBUSB_SAMPLE`
+* :option:`CONFIG_USB_PID_BLE_HCI_H4_SAMPLE`
```
Something is wrong with the commit message as this also adds the driver.

Vudentz replied:
Well it was meant to add the PID with its driver, or are you suggesting this should be splited in 2 commits?

jfischer-phytec-iot replied:
No, but please update the commit title, it somehow does not reflect the changes in the patch self.

Vudentz replied:
Sure, will do update the description.

jfischer-phytec-iot commented on subsys/usb/class/bt_h4.c:
```diff
@@ -41,13 +41,13 @@ static K_FIFO_DEFINE(tx_queue);
 
-#define BT_H4_OUT_EP_ADDR		0x01
-#define BT_H4_IN_EP_ADDR		0x81
+#define BT_H4_OUT_EP_ADDR               0x01
```
Changes here (commit e81d3694161617c1ef29b11ed293ca9627033e03) looks like a rebase accident

Vudentz replied:
Right, this was meant to be a fixup to the previous patch, will fix it.

Vudentz replied:
Fixed.

jfischer-phytec-iot commented:
> can you confirm that you got an HCI Controller with these changes?

@Vudentz tested with reel board (nRF52840), hci_usb seems to work, host also enumerates hci_usb_h4

Vudentz commented:
> > can you confirm that you got an HCI Controller with these changes?
> 
> @Vudentz tested with reel board (nRF52840), hci_usb seems to work, host also enumerates hci_usb_h4

So I guess we are good to go since Ive addressed your comments as well.

emob-nordic commented:
What can I use to use hci_usb_h4 device? hcitool seems not to see it.

jhedberg commented on samples/bluetooth/hci_uart/src/main.c:
```diff
@@ -154,12 +142,18 @@ static void bt_uart_isr(struct device *unused)
-			switch (type) {
-			case H4_CMD:
-				buf = h4_cmd_recv(&remaining);
+			buf = bt_buf_get_tx(BT_BUF_ACL_OUT, K_NO_WAIT, &type,
+					    sizeof(type));
+			if (!buf) {
+				return;
+			}
+
+			switch (bt_buf_get_type(buf)) {
```
This looks broken to me.

Firstly, it looks super odd to pass explicitly `BT_BUF_ACL_OUT` as the first parameter to `bt_buf_get_tx()`. when you don't know what type the packet actually is. Secondly, that function uses this first parameter to call `bt_buf_set_type()`, so you'll never get anything back in the switch statement except `ACL_OUT`.

jhedberg replied:
Actually, `bt_buf_get_tx()` calls `bt_buf_set_type()` twice, the second time overwriting the first, which confused me. Can you update the code so that it only sets the type once?

jhedberg replied:
And add a new `BT_BUF_AUTO` enum for this purpose. Whenever HCI raw H4 isn't enabled passing `BT_BUF_AUTO` should return error, or perhaps better, assert.

Vudentz replied:
Will update the code with BT_BUF_AUTO, could it perhaps be called BT_BUF_H4 which is onlt defined when BT_HCI_RAW_H4 is selected?

Vudentz replied:
Done.

jhedberg commented on subsys/bluetooth/host/hci_raw.c:
```diff
@@ -51,11 +69,86 @@ struct net_buf *bt_buf_get_rx(enum bt_buf_type type, s32_t timeout)
 	struct net_buf *buf;
 
+	switch (type) {
+	case BT_BUF_EVT:
+	case BT_BUF_ACL_IN:
+		break;
+	default:
+		BT_ERR("Invalid type: %u", type);
+		return NULL;
```
Wouldn't you get to this "Invalid type" branch now that you added `BT_BUF_H4`?

Vudentz replied:
Actually BT_BUF_H4 shall not be passed to bt_buf_get_rx since this is used for data originating from the stack itself, so it wont have H:4 header which is actually added later in the function.

jhedberg replied:
Sorry, I thought I was looking at `get_tx()`

jhedberg reviewed (approved):
Looks good now from my side, but it seems there's a conflict that has crept in with `subsys/bluetooth/controller/hci/hci.c`

Vudentz commented:
> Looks good now from my side, but it seems there's a conflict that has crept in with `subsys/bluetooth/controller/hci/hci.c`

It should be resolved now.
